cpp/src/join/join_common_utils.cuh

```diff
@@ -0,0 +1,263 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <join/join_common_utils.hpp>
+
+#include <cudf/detail/utilities/cuda.cuh>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_uvector.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/sequence.h>
+#include <cub/cub.cuh>
+
+namespace cudf {
+namespace detail {
+
+/**
+ * @brief Computes the trivial left join operation for the case when the
+ * right table is empty. In this case all the valid indices of the left table
+ * are returned with their corresponding right indices being set to
+ * JoinNoneValue, i.e. -1.
+ *
+ * @param left Table of left columns to join
+ * @param stream CUDA stream used for device memory operations and kernel launches
+ * @param mr Device memory resource used to allocate the result
+ *
+ * @return Join output indices vector pair
+ */
+inline std::pair<std::unique_ptr<rmm::device_uvector<size_type>>,
+                 std::unique_ptr<rmm::device_uvector<size_type>>>
+get_trivial_left_join_indices(
+  table_view const& left,
+  rmm::cuda_stream_view stream,
+  rmm::mr::device_memory_resource* mr = rmm::mr::get_current_device_resource())
+{
+  auto left_indices = std::make_unique<rmm::device_uvector<size_type>>(left.num_rows(), stream, mr);
+  thrust::sequence(rmm::exec_policy(stream), left_indices->begin(), left_indices->end(), 0);
+  auto right_indices =
+    std::make_unique<rmm::device_uvector<size_type>>(left.num_rows(), stream, mr);
+  thrust::fill(
+    rmm::exec_policy(stream), right_indices->begin(), right_indices->end(), JoinNoneValue);
+  return std::make_pair(std::move(left_indices), std::move(right_indices));
```

Comment:
Not sure it matters, but the fill and sequence could be combined using a zip iterator of a counting iterator and a constant iterator, and then a thrust::copy or thrust::for_each. Then this would be a single kernel.

But really my question is do these really need to be materialized? Could the thing that uses them use iterators instead?

Reply:
These need to be materialized because the output is what is returned to the user when the user requests a left join and there are no common rows between the two tables. I agree, we could fuse the kernels. The current approach seems more readable to me since these are the two primary outputs, and in any case this code path is not performance critical, but I'm not attached to either approach and can change this if you'd like.